### DISCUSSION TITLE
Order of data in FIMS

### DISCUSSION THREAD

**Discussion by kellijohnson-NOAA** [NOAA-FIMS/FIMS#700] [Developers]
Hi <ins>ALL</ins> 😃 I am curious when we are going from a long data frame (i.e., FIMSFrame) to a vector, what order should the data be in? The accessor functions need to ensure the ordering is followed in an appropriate manner but I honestly do not know what that order is 🤦, especially now that we have added length. For example, with age-proportion data, should it be sorted/ordered/arranged first by fleet, year, then age bin? For length data, for example, an age-length-conversion key does it go fleet, year, age bin, and then length bin? We will need to hardcode this order in FIMSFrame or as an internal data object so that `dplyr::arrange()` can be used appropriately before @data is created in the FIMSFrame object. Thank you in advance for participating in this discussion. I am tagging those that I MUST HAVE feedback from but everyone should feel free to comment.
@msupernaw @Bai-Li-NOAA @nathanvaughan-NOAA

**Comment by nathanvaughan-NOAA** [comment]
I don't have any strong opinion on the ordering. We should make sure that it remains consistent with the naming of the indexing variables in the FIMS code though i.e. a vector indexed by i_age_year should be sorted by age within year. I think the best approach would be to have the ordering sorted out by the import/wrapper functions so that the ordering in the long FIMSFrame doesn't matter because it has all the other values to reference. It would be best if we can avoid having users interact with FIMS inputs or outputs in ways that could lead to miss ordering.

**Comment by iantaylor-NOAA** [comment]
I like the suggested "fleet, year, age bin, and then length bin". SS3 doesn't care what order your inputs are in, but that's essentially the order I use all the time anyway because I find it most intuitive and I think it matches the hierarchy in which we process our data, typically one fleet at a time and then the bins are relative to each other within a year / time-step. Also, it's what we're using already for data_mile1 (without yet the length part).

I presume that FIMSFrame could similarly be automatically sorted so the user doesn't get tripped up if they accidentally enter stuff out of order, so this is really just for the internal calculations, right?

**Reply by kellijohnson-NOAA** [reply]
Yes, @iantaylor-NOAA that is what I was thinking, i.e., the ordering would be done in FIMSFrame() so it wouldn't matter what order the user passes their data to FIMSFrame() because it would be done/forced for them.
